sympy/tensor/array/arrayop.py

```diff
@@ -269,6 +269,9 @@ def derive_by_array(expr, dx):
     from sympy.tensor.array import SparseNDimArray
     array_types = (Iterable, MatrixBase, NDimArray)
 
+    expr = sympify(expr)
+    dx = sympify(dx)
+
     if isinstance(dx, array_types):
```

Comment:
Can we just replace this `if` with a `try` that calls the constructor and sees if it works.

Reply:
Or use `_sympify` but just move it below these `if` blocks.

Reply:
> Or use `_sympify` but just move it below these `if` blocks.

Yes it worked. Thanks!!